subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +328,291 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
+			klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+				"than the total free cells of all the VCs (%v). Certain VCs' cells may be doomed to be bad.",
+				h.cellTypes[chain][level], chain, level, h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level],
+				h.badFreeCellNum[chain][level], h.allVCFreeCellNum[chain][level])
+			for vcName, vcQuota := range h.vcFreeCellNum {
+				if vcQuota[chain][level] > h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
+					klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+						"than the free cells of the VC %v (%v). Certain cells in the VC are doomed to be bad.",
+						h.cellTypes[chain][level], chain, level,
+						h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level], h.badFreeCellNum[chain][level],
+						vcName, vcQuota[chain][level])
+					for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[chain][level] {
+						virtualCell := vc.(*VirtualCell)
+						if virtualCell.GetPhysicalCell() == nil && virtualCell.GetAPIStatus().CellHealthiness != api.CellBad {
+							virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+							break
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if parent := c.GetParent(); parent != nil {
+		for _, buddy := range parent.GetChildren() {
+			if buddy.(*PhysicalCell).GetAPIStatus().CellHealthiness != api.CellBad {
+				return
+			}
+		}
+		h.setBadCell(parent.(*PhysicalCell))
+	}
+}
+
+// setHealthyCell marks a physical cell as healthy and recursively for its parent, guaranteeing that
+// a cell is healthy if any of its children is healthy. The virtual cell that this physical cell
+// is bound to will also be marked as healthy.
+func (h *HivedAlgorithm) setHealthyCell(c *PhysicalCell) {
+	klog.Infof("Cell %v is set to healthy", c.GetAddress())
+	c.GetAPIStatus().CellHealthiness = api.CellHealthy
+	if c.virtualCell != nil {
+		c.GetAPIStatus().VirtualCell.CellHealthiness = api.CellHealthy
+		c.virtualCell.GetAPIStatus().CellHealthiness = api.CellHealthy
+		c.virtualCell.GetAPIStatus().PhysicalCell.CellHealthiness = api.CellHealthy
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	if inFreeCellList(c) {
+		h.badFreeCellNum[chain][level]--
+		if h.allVCFreeCellNum[chain][level] >= h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
+			if h.allVCFreeCellNum[chain][level] == h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
+				klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells (healthy %v, bad %v) "+
+					"for allocating the total free cells of all the VCs (%v).",
+					h.cellTypes[chain][level], chain, level, h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level],
+					h.badFreeCellNum[chain][level], h.allVCFreeCellNum[chain][level])
+			}
+			for vcName, vcQuota := range h.vcFreeCellNum {
```

Comment:
>vcQuota [](start = 15, length = 7)

Rename quota to be a more concert concept. Such as total/free cells. Also for all places.